src/pages/2023-12/next-js-app-router-cache/index.mdx

```diff
@@ -0,0 +1,586 @@
+---
+layout: "@layouts/BlogPost.astro"
+title: "Next.js App Router Caching Mechanisms"
+date: "2023-12-15"
+description: "Next.js improves your application's performance and reduces costs by caching rendering work and data requests. This article goes in depth about everything you need to know about Next.js's caching mechanisms."
+tags: ["Next.js"]
+---
+
+import Tangent from "@blogComponents/lib/Tangent.astro";
+
+## Introduction
+
+Next.js is an amazing framework that makes writing complex server rendered React apps much easier, but there is one huge problem. Next.js's caching mechanism is extremely complicated and can easily lead to bugs in your code that are difficult to debug and fix.
+
+If you don't understand how Next.js's caching mechanism works it feels like you are constantly fighting Next.js instead of reaping the amazing benefits of Next.js's powerful caching. That is why in this article I am going to break down exactly how every part of Next.js's cache works so you can stop fighting it and finally take advantage of its incredible performance gains.
+
+If you are not already familiar with React Server Components, I wrote an entire article on it which explains them in depth. You can read that [here](https://blog.webdevsimplified.com/2023-11/react-server-components/).
+
+Here is an image of how all the caches in Next.js interact with one another. I know this is overwhelming, but by the end of this article you will understand exactly what each step in this process does and how they all interact.
+
+![cache-interactions](/articleAssets/2023-12/next-js-app-router-cache/cache-interactions.png)
+
+## Next.js Caching Mechanisms
+
+The main reason Next.js' cache is so hard to understand is because it is
+actually broken down into 4 distinct caching mechanisms that all operate on different stages of your application and interact with each other in complicated ways.
+
+The 4 caching mechanisms are:
+
+1. Request Memoization
+2. Data Cache
+3. Full Route Cache
+4. Router Cache
+
+I will be covering each cache's purpose, where it is stored, how long it lasts for, how to invalidate the cache and how to opt out. This seems like a lot of information at first glance but we will take an example based approach.
```

Comment:
This is missing a comma before `and`. Also you can remove the last sentence from this as well.